### PR TITLE
Implement FusedAdam for ORT adapted from DeepSpeed

### PR DESCRIPTION
**Pull request by baijumeswani**

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cuda/adam_optimizer/type_shim.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            using scalar_t_##LEVEL = float;                                      \
+            __VA_ARGS__;                                                         \
+            break;                                                               \
+        }                                                                        \
+        default: AT_ERROR(#NAME, " not implemented for '", toString(TYPE), "'"); \
+    }
+
+template <typename T>
+__device__ __forceinline__ T
+reduce_block_into_lanes(T* x,
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cuda/adam_optimizer/type_shim.h:73]: this function is not used anywhere. can remove.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -3977,3 +3979,32 @@ def __init__(self, module, debug_options=None):
...
+        return loss
+
+    def run_optim_step(optimizer):
+        optimizer.step()
+        optimizer.zero_grad()
+
+    ga_steps = 2
+    ort_model.zero_grad()
+
+    for step in range(10):
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:4005]: This can be done in the next PR.
take one pytorch model, run with transformer.AdamW 
Take the cloned ortmodule model, run with ort.fusedadam. 

Compare model.weight, param.grad, loss for every steps. assert the values is close.

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cuda/adam_optimizer/compat.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+/* Copyright 2020 The Microsoft DeepSpeed Team
+   Copyright NVIDIA/apex
+   This file is adapted from fused adam in NVIDIA/apex, commit a109f85
+*/
+
+#ifndef TORCH_CHECK
+#define TORCH_CHECK AT_CHECK
+#endif
+
+#ifdef VERSION_GE_1_3
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cuda/adam_optimizer/compat.h:14]: This macro won't work, since we didn't have this macro defined anywhere. 

However, I don't think we would need this macro either, as we will only support torch>1.8. 

Just replace the use of "DATA_PTR" with  "data"

Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cuda/adam_optimizer/compat.h:14]: remove this head file, and all the #include "compat.h" occurrences.

**Review by SherlockNoMad — DISMISSED** [review]

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:4010]: Updated the unit test to do this.
